plugin/call_hierarchy.py

```diff
@@ -0,0 +1,168 @@
+            element['name'],
+            kind=SYMBOL_KINDS.get(element['kind'], sublime.KIND_AMBIGUOUS),
+            description=element.get('detail', ""),
+            tooltip="{}:{}".format(self._simple_path(element['uri']), element['selectionRange']['start']['line'] + 1),
+            command_url=command_url
+        )
+
+    def _simple_path(self, uri: DocumentUri) -> str:
+        scheme, path = parse_uri(uri)
+        session = self.weaksession()
+        if not session or scheme != 'file':
+            return path
+        simple_path = simple_project_path([Path(folder.path) for folder in session.get_workspace_folders()], Path(path))
+        return str(simple_path) if simple_path else path
+
+    def _handle_incoming_calls_async(
+        self, response: Optional[List[CallHierarchyIncomingCall]]
+    ) -> List[CallHierarchyItem]:
+        return [incoming_call['from'] for incoming_call in response] if response else []
+
+    def _handle_outgoing_calls_async(
+        self, response: Optional[List[CallHierarchyOutgoingCall]]
+    ) -> List[CallHierarchyItem]:
+        return [outgoing_call['to'] for outgoing_call in response] if response else []
+
+
+class LspCallHierarchyCommand(LspTextCommand):
+
+    capability = 'callHierarchyProvider'
+
+    def is_visible(self, event: Optional[dict] = None, point: Optional[int] = None) -> bool:
+        if self.applies_to_context_menu(event):
+            return self.is_enabled(event, point)
+        return True
+
+    def run(self, edit: sublime.Edit, event: Optional[dict] = None, point: Optional[int] = None) -> None:
+        self._window = self.view.window()
+        session = self.best_session(self.capability)
+        if not session:
+            return
+        position = get_position(self.view, event, point)
+        if position is None:
+            return
+        params = cast(CallHierarchyPrepareParams, text_document_position_params(self.view, position))
+        request = Request.prepareCallHierarchy(params, self.view)
+        session.send_request_async(request, partial(self._handle_response_async, weakref.ref(session)))
+
+    def _handle_response_async(
+        self, weaksession: 'weakref.ref[Session]', response: Optional[List[CallHierarchyItem]]
+    ) -> None:
+        if not self._window or not self._window.is_valid():
+            return
+        if not response:
+            self._window.status_message("Call hierarchy not available")
+            return
+        session = weaksession()
+        if not session:
+            return
+        data_provider = CallHierarchyDataProvider(weaksession, CallHierarchyDirection.IncomingCalls, response)
+        header = 'Call Hierarchy: Callers of… {}'.format(
+            make_command_link('lsp_call_hierarchy_toggle', "⇄", {
+                'session_name': session.config.name,
+                'direction': CallHierarchyDirection.OutgoingCalls,
+                'root_elements': response
+            }, tooltip="Show outgoing calls"))
```

Comment:
Is it intended that clicking this opens a new sheet? I think that's not how it worked when I've tested it initially.

Reply:
Actually that only happens after restarting ST and interacting with pre-existing sheet.

Reply:
The `new_tree_view_sheet` function, which is called when you click on this, should ensure that there is always only one sheet with a particular name (e.g. "Call Hierarchy", "Type Hierarchy", etc.) per window, and the sheet is stored in the `WindowManager`. But yeah, if you restart ST, then this data isn't preserved and a new `WindowManager` is initialized with empty data. I didn't run into this because I use `"hot_exit": "disabled"`.

Also interaction with the tree view doesn't work anymore if you drag the tab out of the window, for the same reason. But I think this is a broader issue in general, and various other LSP features suffer from the same problem. This is mainly a fundamental problem of the idea to associate "workspace folders" or "sessions" to a particular window, and it probably can't be solved completely.

Reply:
I don't mind this issue for now.
Ideally the sheet would not survive the restart IMO but I suppose there is no easy way to achieve that right now.